### DISCUSSION TITLE
The garbage collector is trying to clean up connection...

### DISCUSSION THREAD

**Discussion by connorharv** [sqlalchemy/sqlalchemy#6416] [Usage Questions]
I'm running a discord bot (https://github.com/Conbonbot/RS-Club-Bot) and one of the features of it is it checks the time every minute to see if someone has been in a queue for over 60 minutes, but I keep getting the error:
```
2021-04-28 11:20:26,284 INFO sqlalchemy.engine.Engine [cached since 30.71s ago] (837030385829150720,)
INFO:sqlalchemy.engine.Engine:[cached since 30.71s ago] (837030385829150720,)
ERROR:sqlalchemy.pool.impl.AsyncAdaptedQueuePool:The garbage collector is trying to clean up connection <sqlalchemy.dialects.postgresql.asyncpg.AsyncAdapt_asyncpg_connection object at 0x7fc0a2113f40>. This feature is unsupported on async dbapi, since no IO can be performed at this stage to reset the connection. Please close out all connections when they are no longer used, calling ``close()`` or using a context manager to manage their lifetime.
/home/conbonbot/rs_bot/env/lib/python3.8/site-packages/sqlalchemy/pool/base.py:693: SAWarning: The garbage collector is trying to clean up connection <sqlalchemy.dialects.postgresql.asyncpg.AsyncAdapt_asyncpg_connection object at 0x7fc0a2113f40>. This feature is unsupported on async dbapi, since no IO can be performed at this stage to reset the connection. Please close out all connections when they are no longer used, calling ``close()`` or using a context manager to manage their lifetime.
  util.warn(message)
ERROR:sqlalchemy.pool.impl.AsyncAdaptedQueuePool:The garbage collector is trying to clean up connection <sqlalchemy.dialects.postgresql.asyncpg.AsyncAdapt_asyncpg_connection object at 0x7fc0a32512c0>. This feature is unsupported on async dbapi, since no IO can be performed at this stage to reset the connection. Please close out all connections when they are no longer used, calling ``close()`` or using a context manager to manage their lifetime.
2021-04-28 11:20:26,308 INFO sqlalchemy.engine.Engine BEGIN (implicit)
/home/conbonbot/rs_bot/env/lib/python3.8/site-packages/sqlalchemy/pool/base.py:693: SAWarning: The garbage collector is trying to clean up connection <sqlalchemy.dialects.postgresql.asyncpg.AsyncAdapt_asyncpg_connection object at 0x7fc0a32512c0>. This feature is unsupported on async dbapi, since no IO can be performed at this stage to reset the connection. Please close out all connections when they are no longer used, calling ``close()`` or using a context manager to manage their lifetime.
  util.warn(message)
```

The code that checks the time every minute is:

```python
@tasks.loop(minutes=1.0)
    async def check_people(self):
        # This command will run every minute, and check if someone has been in a queue for over n minutes
        LOGGER.debug("Attempting to check the time")
        session = sessionmaker()
        results = (await session.execute(select(Queue))).scalars()
        for queue in results:
            minutes = int((time.time() - queue.time) / 60)
            if minutes == queue.length:
                # Ping the user
                    user = await self.bot.fetch_user(queue.user_id)
                    channel = await self.bot.fetch_channel(queue.channel_id)
                    message = await channel.send(
                        f"{user.mention}, still in for a RS{queue.level}? React ✅ to stay in the queue, and ❌ to leave the queue")
                    await message.add_reaction('✅')
                    await message.add_reaction('❌')
                    # Add their user_id and message_id to database
                    add_temp = Temp(server_id=queue.server_id, user_id=queue.user_id, message_id=message.id, amount=queue.amount, level=queue.level)
                    session.add(add_temp)
                    await session.commit()

                    ctx = await self.bot.get_context(message.id)
                    await self.print_queue(ctx, self.rs_channel[str(ctx.message.channel)])
            elif minutes >= queue.length + 5:
                    User_leave = (await session.get(Queue, (queue.server_id, queue.user_id, queue.amount, queue.level)))
                    await session.delete(User_leave)
                    await session.commit()
                    user = await self.bot.fetch_user(queue.user_id)
                    channel = await self.bot.fetch_channel(queue.channel_id)
                    await channel.send(f"{user.display_name} has left RS{queue.level} ({await self.amount(queue.level)}/4)")
                    conditions = and_(Temp.user_id == queue.user_id, Temp.level == queue.level)
                    id = (await session.execute(select(Temp).where(conditions))).scalars()
                    for i in id:
                        message = await channel.fetch_message(i.message_id)
                        await message.delete()
                    conditions = and_(Temp.server_id == queue.server_id, Temp.user_id == queue.user_id, Temp.level == queue.level)
                    await session.execute(delete(Temp).where(conditions))
                    await session.commit()
        await session.commit()
        await session.close()
```

I do have other commands that use SQLAlchemy's sessions, like this:
```python

    @commands.command(aliases=["r"])
    async def refresh(self, ctx):
        session = sessionmaker()
        conditions = and_(Queue.user_id == ctx.author.id, level=self.rs_channel[str(ctx.message.channel)])
        times = (await session.execute(select(Queue).where(conditions))).scalars()
        times.first().time = int(time.time())
        rs_level = f'RS{self.rs_channel[str(ctx.message.channel)]}'
        num_players = f'({await self.amount(self.rs_channel[str(ctx.message.channel)])}/4)'
        await ctx.send(f"{ctx.author.mention}, you are requeued for a {rs_level}! {num_players}")
        await session.commit()
        await session.close()
```
But in every case I do commit and close each session when I'm done with them.

Most of the code involving SQLAlchemy is in this file: https://github.com/Conbonbot/RS-Club-Bot/blob/master/routines/cogs/queue.py

Any help on this would be greatly appreciated!

**Comment by CaselIT** [comment]
Hi,

It's probably hard to understand from where this warning is coming from since the usual python flag `-Werror` may not help.
Quickly looking at the code it seems that there are places where the session are not closed. Like
https://github.com/Conbonbot/RS-Club-Bot/blob/a39ad13a15ceed59b23aa49efb9ef7da169da899/routines/cogs/queue.py#L231-L239
or
https://github.com/Conbonbot/RS-Club-Bot/blob/a39ad13a15ceed59b23aa49efb9ef7da169da899/routines/cogs/queue.py#L352-L473

Also note that in the example you provided the session would not be close in case of errors.
I suggest that you instead use a context manager instead of manually closing the sessions:
```py
async with sessionmaker() as session:
  ...
```

**Reply by connorharv** [reply]
Thanks for the help, I thought I manually closed every session when I was done with them, but I'll use a context manager like you mentioned, it could be that something else is breaking that prevents the sessions from closing.

**Reply by connorharv** [reply]
I've changed everything to use the code you mentioned above and I let it run for a couple days and I'm still getting the error about the garbage collector, I think it could be an issue with
```
    @tasks.loop(minutes=1.0)
    async def check_people(self):
```
and how @tasks.loop actually works.

**Reply by zzzeek** [reply]
if Loop (seen here: https://github.com/Rapptz/discord.py/blob/master/discord/ext/tasks/__init__.py) is pre-empting the routine, it would still be doing that using an exception raise of some kind that a "finally" block or equivalent should catch in order to do clean up.   but then I dont know how "finally:" that has awaitable cleanup logic inside of it works, if the awaitable is being interrupted.

**Reply by zzzeek** [reply]
I thnk you need to add more logging to that routine at least to se what's happening

**Reply by connorharv** [reply]
Yeah I think you're right that @tasks.loop is pre-empting the routine. I'll try to isolate the issue and add more logging to find out more about this issue, and thank you for your response.

**Reply by CaselIT** [reply]
But even if pre-empting the routine I would think the aexit should still be called

**Reply by zzzeek** [reply]
yeah but if you have non blocking IO in atexit or whatever, asyncio event loop is done or how does that work

**Comment by zzzeek** [comment]
so this is not the "answer", @CaselIT 's advice is pretty much where we need to be at the moment, but I have comments !

1. im super psyched to see the warning working, and note that we just drop the connection at that phase, where I would *assume*, but I haven't confirmed, that from an asycnpg perspective the actual PG session would be closed out in any case, but we might want to check the asyncpg docs.  like what happens in asyncpg if you just dereference a connection, does the PG session get closed  ?    it should, ultimately this is just a socket with O_NONBLOCK set on it.   So the warnings here are *probably* not implying anything more than something is not managing to close out every connection
2. super psyched to see fully 2.0 -style SQLAlchemy ORM code !!   the 2.0 style API, most notably the "session.execute()" thing, was largely pushed by the asyncio idea of "make it explicit where IO happens".   I think it looks great
3. more specifically what does `@tasks.loop` do here, is there any chance it is pre-empting the function's execution, or perhaps running it more than once on top of itself if it runs slowly, either of which could interrupt the process?
4. im beginning to think of ways that this kind of thing could be traced more accurately, like a connection pool that memoizes the stack trace of where the checkout occurred, so that when we get the warning, it dumps it out.  I think if we add an event hook for this case, we can have a recipe that does this.   right now there's an event hook `close_detached` that gets called in the case that we *can* close this connection, so maybe another event `abandon_detached`, where we document this is pretty much just for asyncio and it's pretty much for this case.    then one can create a `checkout` handler that stores stack traces and an `abandon_detached` / `close_detached` handler that logs where these "unhandled" connections were checked out.   I think we might have done something like this a long time ago.
